### ISSUE TITLE
Custom encryption

### ISSUE THREAD

**Issue by ManimaranNCOG** [dbmail/dbmail#278]
I am integrating Postfix with DBMail (using PostgreSQL). Currently, when an email is received by Postfix, DBMail encodes the email using the default UTF-8 encoding to store it in PostgreSQL. However, I need to use custom encryption to store the emails in the database. Can you provide guidance on implementing custom encryption for storing emails in the database?

**Comment by alan-hicks** [comment]
There are two config values that may be helpful
[DBMAIL]
# 
# encoding must match the database/table encoding.
# i.e. latin1, utf8
encoding             = utf8

#
# messages with unknown encoding will be assumed to have 
# default_msg_encoding
# i.e. iso8859-1, utf8
default_msg_encoding = utf8

Hope these help

**Comment by ManimaranNCOG** [comment]
Thanks for your comment.

We're looking to encrypting emails using users' unique public keys for storage in a database. Each user, identified by their email address like 123@mydomain.com and abcd@mydomain.com, has their own unique public key. The goal is to encrypt emails using these public keys before storing them in the database. If there's a way to achieve this encryption, please let me know.

Thanks.

**Comment by CozC** [comment]
Salve,

Well, short answer is no, dbmail does not posses such a feature.

Long answer, it needs some work, but it can be done. Basically DBMail  
stores the email in a few tables, and those are split in 3, tables that  
holds information about the metadata, tables that holds the headers and  
tables that holds the body part of the message.
Basically one needs to intercept those functions and apply the required  
encryption.  So you need to decide which part of the those 3, you need to  
be encrypted. My take is that you are going for the body part and maybe  
the headers which are implemented in 2 tables, so you need to redirect the  
methods(functions) that do retrieving and storing the information.
First you may need to develop a storage of public private keys (same db or  
some other type of external storage, may a (soft)HSM). You may need to  
think that the encoding of large volumes of data with private/public keys  
is slow so you may have to think about a schema, use the asymmetric part  
to decrypt a key to be used in a symmetric encoding decoding.

So bottom line... it can be done.

The second idea, it might not be exactly what you are searching for but  
for the sake of argument. You can employ database encryption and by  
employing strong measures to access your box (nodes) it should be enough  
for the security perspective.

Hope it helps.
-- 
CC @gmail

**Comment by alan-hicks** [comment]
Depending on what you decide to encrypt, you might want to disable 5.1.2. Shared Namespace (https://datatracker.ietf.org/doc/html/rfc9051)
Also there is a lot of caching (https://dbmail.org/architecture/email-cache/) so please test your solution.

There may be easier alternatives including https://en.wikipedia.org/wiki/GNU_Privacy_Guard and https://en.wikipedia.org/wiki/S/MIME

**Comment by ManimaranNCOG** [comment]
Thank you for the valuable information!

**Comment by alan-hicks** [comment]
Closing as no feedback for a long time.
